### PR TITLE
Account api tests

### PR DESCRIPTION
**Pull request by chrisveilleux**
## Description
Improve upon the account API behave tests.

## How to test
Run the account API behave tests.  All should pass.

## Contributor license agreement signed?
CLA [Y]

### CODE REVIEW

File: api/account/tests/features/steps/profile.py [hunk 1]
```diff
@@ -18,67 +18,87 @@
...
+@given("an account without a membership")
 def get_account_no_membership(context):
-    context.username = 'foo'
+    context.username = "foo"
+
 
+@given("an account opted {in_or_out} the Open Dataset agreement")
+def set_account_open_dataset(context, in_or_out):
+    context.username = "foo"
+    if in_or_out == "out of":
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:77]: The environment.py file builds the account with opt-in by default.

File: api/account/tests/features/steps/profile.py [hunk 2]
```diff
@@ -87,84 +107,128 @@ def add_membership_via_api(context):
...
-        equal_to('Monthly Membership')
-    )
-    assert_that(response_data['membership']['duration'], none())
-    assert_that(
-        response_data['membership'], has_item('id')
-    )
+    utc_date = datetime.utcnow().date()
+    account = context.accounts["foo"]
+    assert_that(response_data["emailAddress"], equal_to(account.email_address))
+    assert_that(response_data["membership"]["type"], equal_to("Monthly Membership"))
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:144]: "Monthly Membership" is the default value set in environment.py, so it should always be this value.

File: api/account/tests/features/steps/profile.py [hunk 3]
```diff
@@ -87,84 +107,128 @@ def add_membership_via_api(context):
...
-    assert_that(len(response_data['agreements']), equal_to(3))
-    agreement = response_data['agreements'][0]
-    assert_that(agreement['type'], equal_to(PRIVACY_POLICY))
-    assert_that(
-        agreement['acceptDate'],
-        equal_to(str(date.today().strftime('%B %d, %Y')))
-    )
-    assert_that(agreement, has_item('id'))
+    assert_that(len(response_data["agreements"]), equal_to(3))
+    agreement = response_data["agreements"][0]
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:149]: Good idea.  Changing to for loop to check all agreements

File: api/account/tests/features/steps/profile.py [hunk 4]
```diff
@@ -87,84 +107,128 @@ def add_membership_via_api(context):
...
+    acct_activity_repository = AccountActivityRepository(context.db)
+    account_activity = acct_activity_repository.get_activity_by_date(
+        datetime.utcnow().date()
+    )
+    if context.account_activity is None:
+        assert_that(account_activity.members, greater_than(0))
+        assert_that(account_activity.members_expired, equal_to(1))
+    else:
+        # Membership was added in a previous step so rather than the membership being
+        # decreased by one, it would net to being the same after the expiration.
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:204]: do you mean change "net" to "need"?  if so, I did intend to use the word "net".  can improve wording if this is confusing.

File: api/account/tests/features/steps/profile.py [hunk 5]
```diff
@@ -87,84 +107,128 @@ def add_membership_via_api(context):
...
+        )
+
+
+@then("the account {will_or_wont} have a open dataset agreement")
+def check_for_open_dataset_agreement(context, will_or_wont):
+    account_repo = AccountRepository(context.db)
+    account = account_repo.get_account_by_id(context.accounts["foo"].id)
+    agreements = [agreement.type for agreement in account.agreements]
+    if will_or_wont == "will":
+        assert_that("Open Dataset", is_in(agreements))
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:220]: nice catch!

File: api/account/tests/features/steps/profile.py [hunk 6]
```diff
@@ -87,84 +107,128 @@ def add_membership_via_api(context):
...
+    else:
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/account/tests/features/steps/profile.py:221]: probably just me being lazy.  fixing

File: api/public/tests/features/device_metrics.feature [hunk 7]
```diff
@@ -1,14 +1,15 @@
 Feature: Save metrics sent to selene from mycroft core
 
-  Scenario: Metric sent by device saved to database
-    Given an existing device
-    When the metric endpoint is called
-    Then the metric is saved to the database
-    And the request will be successful
+  Scenario: User opted into the open dataset uses their device
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/public/tests/features/device_metrics.feature:3]: we won't get device metrics from an opted-out user.

File: api/public/tests/features/steps/device_metrics.py [hunk 8]
```diff
@@ -19,51 +19,50 @@
...
+METRIC_TYPE_TIMING = "timing"
+metric_value = dict(type="timing", start="123")
 
 
-@given('an existing device')
-def define_authorized_device(context):
-    context.metric_device_id = context.device_login['uuid']
+@given("a device registered to a user opted {in_or_out} the open dataset")
+def define_authorized_device(context, in_or_out):
+    context.metric_device_id = context.device_login["uuid"]
```
[the remark this answers is not in the document]
Comment (chrisveilleux) [reply api/public/tests/features/steps/device_metrics.py:36]: correct, but I didn't want to have to generate two different step functions that did the exact same thing.

**Comment by chrisveilleux** [comment]
Funny story, this PR was accidentally built to merge into master.  Changed to merge into dev and all the changes went away.  Probably part of an already merged PR.  Anyway, your code review points were valid so now this PR just contains fixes based on your comments.

File: api/account/tests/features/steps/profile.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (krisgesling) [reply api/account/tests/features/steps/profile.py]: Oh yeah that makes sense, nevermind.

File: api/sso/tests/features/steps/password_change.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+
+@given('a user who authenticates with a password')
+def setup_user(context):
+    acct = context.accounts['foobar']
+    context.email = acct.email_address
+    context.password = 'bar'
+    context.change_password_request = dict(
+        accountId=acct.id, password=b2a_base64(b'bar').decode()
+    )
```
[the remark this answers is not in the document]
Comment (krisgesling) [reply api/sso/tests/features/steps/password_change.py]: This is no longer related to this PR but my read of it is that it's changing the password from 'bar' to 'bar', should we actually change the password or have I misinterpreted?

**Pull request merged**